starknet_devnet/util.py

```diff
@@ -263,3 +264,62 @@ def __exit__(self, exc_type, exc_val, exc_tb):
 # FeederGatewayClient is implemented in such a way that it logs and raises;
 # this suppresses the logging
 suppress_feeder_gateway_client_logger = LogSuppressor("services.external_api.client")
+
+logger = logging.getLogger("gunicorn.error")
+
+
+class LogContext(dict):
+    """Context manager for logging transactions accumulating info while executing transaction"""
+
+    def get_context_name(self) -> str:
+        """Getting name of current context to put as first line in log entry"""
+        return getattr(self, "name", None) or "Transaction"
+
+    def set_context_name(self, name):
+        """Setting name of current context to put as first line in log entry"""
+        setattr(self, "name", name)
+        return self
+
+    def cust_fees_to_int(self):
+        """Cust to human readable type"""
+        for key in ["actual_fee", "max_fee"]:
+            if self.get(key):
+                try:
+                    self[key] = int(self[key], 16)
+                except (ValueError, TypeError):
+                    pass
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.cust_fees_to_int()
+        if exc_type:
+            if isinstance(exc_val, StarkException):
+                error_message = {
+                    "message": exc_val.message,
+                    "code": str(exc_val.code),
+                    "context": self,
+                }
+            else:
+                error_message = {"message": str(exc_val), "context": self}
+            logger.error(
+                "%s failed: %s", self.get_context_name(), pprint.pformat(error_message)
+            )
+            return
+        logger.info("%s completed: %s", self.get_context_name(), pprint.pformat(self))
+
+
+def extract_transaction_info_to_log(transaction: dict) -> dict:
```

Comment:
Why is this necessary? Why not just print the object? It is called the "verbose" mode? Can the output be huge?

Reply:
Sure, made it more verbose

Reply:
I meant my last question (Can the output be huge?) more like: did you reduce the number of printed lines because we are risking the output to be huge and unreadable?

Reply:
I decided to reduce it because I believed that the user wouldn't require such detailed information, only concise "INFO" entries, which could have been extended with context. But in the end, we don't need to keep it at a small size.

With the full verbose mode, it remains readable, except for some contracts that mainly consist of numbers, and I'm unsure if the user would be able to understand them.

Here an example of size deploy transaction with full output: 
```
[2023-06-08 10:06:22 +0000] [5354] [INFO] add_transaction request: {'class_hash': '0x4d07e40e93398ed3c76981e72dd1fd22557a78ce36c0515f679e27f0bb5bc5f',
 'constructor_calldata': ['0x4024999b9574cb7623679ce049a609db62a95098982c5b28ac61abdebd1c82b'],
 'contract_address_salt': '0x99',
 'max_fee': '0xde0b6b3a7640000',
 'nonce': '0x0',
 'signature': ['0x15dc17a076cea6ec2355cde875c7f46b0ff5f312b418afed547f418622b9298',
               '0x246e0207bb454a967abc0b336f6705d7a521017a119d05953cbc25d2e04d4e1'],
 'type': 'DEPLOY_ACCOUNT',
 'version': '0x1'}
[2023-06-08 10:06:23 +0000] [5354] [INFO] transaction execution info: {'actual_fee': '0x64a7168300',
 'actual_resources': {'ecdsa_builtin': 1,
                      'l1_gas_usage': 4284,
                      'n_steps': 3818,
                      'pedersen_builtin': 23,
                      'range_check_builtin': 83},
 'call_info': {'accessed_storage_keys': ['0x1379ac0624b939ceb9dede92211d7db5ee174fe28be72245b0a1a2abd81c98f'],
               'call_type': 'CALL',
               'calldata': [1813293060336465068570802339647185657676088914728234993269889516456358561835],
               'caller_address': 0,
               'class_hash': '0x4d07e40e93398ed3c76981e72dd1fd22557a78ce36c0515f679e27f0bb5bc5f',
               'code_address': None,
               'contract_address': 1065893459572464416143945688077233223288051643985551879776468107551438372322,
               'entry_point_selector': 1159040026212278395030414237414753050475174923702621880048416706425641521556,
               'entry_point_type': 'CONSTRUCTOR',
               'events': [],
               'execution_resources': {'builtin_instance_counter': {},
                                       'n_memory_holes': 0,
                                       'n_steps': 48},
               'failure_flag': '0x0',
               'gas_consumed': '0x0',
               'internal_calls': [],
               'l2_to_l1_messages': [],
               'retdata': [],
               'storage_read_values': [0]},
 'fee_transfer_info': {'accessed_storage_keys': ['0x249c2e4c4e43fd9200796b7abc7f94d3840e2cb6af5cbfba2526ff1c8c28729',
                                                 '0x249c2e4c4e43fd9200796b7abc7f94d3840e2cb6af5cbfba2526ff1c8c2872a',
                                                 '0x43d06bc782c9584c9b9cae590f47bc51fc5ddcbb23d2c0cbe553b86c8f0e040',
                                                 '0x43d06bc782c9584c9b9cae590f47bc51fc5ddcbb23d2c0cbe553b86c8f0e041'],
                       'call_type': 'CALL',
                       'calldata': [1815204096841571598366742751903307880575707433299601255915223373127742545712,
                                    432300000000,
                                    0],
                       'caller_address': 1065893459572464416143945688077233223288051643985551879776468107551438372322,
                       'class_hash': '0x6a22bf63c7bc07effa39a25dfbd21523d211db0100a0afd054d172b81840eaf',
                       'code_address': None,
                       'contract_address': 2087021424722619777119509474943472645767659996348769578120564519014510906823,
                       'entry_point_selector': 232670485425082704932579856502088130646006032362877466777181098476241604910,
                       'entry_point_type': 'EXTERNAL',
                       'events': [{'data': ['0x25b4639c321f3538c69537502f0379a23d6c91d6bf0f9dfe79fabfc3da97de2',
                                            '0x4035e7e01f09b8bc746a549fdf1bafa8d975bdbc53d25646635d0ab58425f30',
                                            '0x64a7168300',
                                            '0x0'],
                                   'keys': ['0x99cd8bde557814842a3121e8ddfd433a539b8c9f14bf31ebf108d12e6196e9'],
                                   'order': 0}],
                       'execution_resources': {'builtin_instance_counter': {'pedersen_builtin': 4,
                                                                            'range_check_builtin': 29},
                                               'n_memory_holes': 42,
                                               'n_steps': 578},
                       'failure_flag': '0x0',
                       'gas_consumed': '0x0',
                       'internal_calls': [],
                       'l2_to_l1_messages': [],
                       'retdata': [1],
                       'storage_read_values': [1000000000000000000,
                                               0,
                                               1000000000000000000,
                                               0,
                                               495000000000,
                                               0,
                                               495000000000,
                                               0]},
 'tx_type': 'DEPLOY_ACCOUNT',
 'validate_info': {'accessed_storage_keys': ['0x1379ac0624b939ceb9dede92211d7db5ee174fe28be72245b0a1a2abd81c98f'],
                   'call_type': 'CALL',
                   'calldata': [2177626953842241688342849477958144159584215338353441986251810208376165874783,
                                153,
                                1813293060336465068570802339647185657676088914728234993269889516456358561835],
                   'caller_address': 0,
                   'class_hash': '0x4d07e40e93398ed3c76981e72dd1fd22557a78ce36c0515f679e27f0bb5bc5f',
                   'code_address': None,
                   'contract_address': 1065893459572464416143945688077233223288051643985551879776468107551438372322,
                   'entry_point_selector': 1554466106298962091002569854891683800203193677547440645928814916929210362005,
                   'entry_point_type': 'EXTERNAL',
                   'events': [],
                   'execution_resources': {'builtin_instance_counter': {'ecdsa_builtin': 1},
                                           'n_memory_holes': 0,
                                           'n_steps': 80},
                   'failure_flag': '0x0',
                   'gas_consumed': '0x0',
                   'internal_calls': [],
                   'l2_to_l1_messages': [],
                   'retdata': [],
                   'storage_read_values': [1813293060336465068570802339647185657676088914728234993269889516456358561835]}}
[2023-06-08 10:06:23 +0000] [5354] [INFO] transaction receipt: {'actual_fee': '0x64a7168300',
 'block_hash': '0x11c4ef8d04df9d839e90f13f43b8f8b87be1057154674cfe875b6673c878845',
 'block_number': 1,
 'events': [{'data': ['0x25b4639c321f3538c69537502f0379a23d6c91d6bf0f9dfe79fabfc3da97de2',
                      '0x4035e7e01f09b8bc746a549fdf1bafa8d975bdbc53d25646635d0ab58425f30',
                      '0x64a7168300',
                      '0x0'],
             'from_address': '0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7',
             'keys': ['0x99cd8bde557814842a3121e8ddfd433a539b8c9f14bf31ebf108d12e6196e9']}],
 'execution_resources': {'builtin_instance_counter': {},
                         'n_memory_holes': 0,
                         'n_steps': 48},
 'l2_to_l1_messages': [],
 'status': 'ACCEPTED_ON_L2',
 'transaction_hash': '0x58a5b426cbda66d208fc300d74e6f371f414feb27f7ff0897efba04b563a5fb',
 'transaction_index': 0}
[2023-06-08 10:06:23 +0000] [5354] [INFO] add_transaction response: {'address': '0x025b4639c321f3538c69537502f0379a23d6c91d6bf0f9dfe79fabfc3da97de2',
 'code': 'TRANSACTION_RECEIVED',
 'transaction_hash': '0x58a5b426cbda66d208fc300d74e6f371f414feb27f7ff0897efba04b563a5fb'}
[2023-06-08 10:06:23 +0000] [5354] [INFO] 127.0.0.1 - - [08/Jun/2023:10:06:23 +0000] "POST /gateway/add_transaction HTTP/1.1" 200 198 "-" "python-requests/2.28.2"
```